posts/nix-vs-docker-i.md

```diff
@@ -0,0 +1,260 @@
+
+```nix
+pkgs.mkShell {
+  buildInputs = [ pkgs.nodejs ];
+  shellHook = ''
+    export NODE_PATH=$(npm config get prefix)/lib/node_modules
+  '';
+}
+```
+
+In a real build derivation, you'd pin the exact Node version and even vendor
+`node_modules` to guarantee purity. That's not just possible---it's the default
+behavior. The build is a pure function. Run it anywhere with the same inputs,
+get the same output.
+
+Nix expressions are declarative. You describe what you want, not how to get
+there. Every package build is a function of its inputs: the source code, the
+compiler, the dependencies, the environment variables, and the build scripts.
+Nothing more. There's no hidden context. No reliance on the global state of your
+machine. No network access during builds unless you explicitly allow it, e.g.,
+with `--no-sandbox`.
+
+## Filesystem and Image Composition
+
+Docker builds up layers. Each command creates a filesystem diff. These are
+cached and stacked. But unless you're extremely careful, you accumulate garbage.
+Consider a typical pattern:
+
+```dockerfile
+RUN apt-get update && apt-get install -y build-essential python3 \
+    && rm -rf /var/lib/apt/lists/*
+```
+
+This trick is needed to keep images small. If you forget to clean up, that layer
+includes every downloaded `.deb` file and package index. There are also
+situations where `/var/lib/apt/lists/` _needs_ to remain in place.
+
+With Nix, you never need these hacks. The build output contains only what was
+declared. For example, here's a `default.nix` to build a Python application:
+
+```nix
+# XXX: The `import <nixpkgs>` pattern is actually discouraged, and it is usually
+# not very reproducible. This is here as an example, but I encourage you to take
+# a look at the twin article for more correct examples with better explanations.
+{ pkgs ? import <nixpkgs> {} }:
+
+pkgs.buildPythonPackage {
+  pname = "myapp";
+  version = "0.1.0";
+  src = ./.;
+  propagatedBuildInputs = [ pkgs.python3Packages.requests ];
+}
+```
+
+No accidental state. No garbage left in `/var`. No need to squash layers. You
+get a single, deduplicated output path like:
+
+```sh
+/nix/store/some-hash-myapp-0.1.0/
+```
+
+That path is input-addressed and reproducible. You can copy it, cache it, or
+reproduce it anywhere. You can even `nix bundle` it to allow running it
+elsewhere, though the API around it is a bit ambiguous and besides the scope of
+today's post.
+
+## Reproducibility
+
+Docker can be made reproducible if you treat it like a hostile system and strip
+away everything non-deterministic. This _is_ possible, but you'll notice very
+quickly that the defaults work against you. Correctness is possible, but it is
+pushed back in the name of 'convenience' for mediocre solutions to difficult
+problems. You can, for example, pin your base image. Vendor your dependencies.
+Disable all dynamic package resolution. Even then, something can—and most likely
+will---still slip through.
```

Comment:
There's also the rabbit hole of filesystem timestamps https://scribe.rip/nttlabs/bit-for-bit-reproducible-builds-with-dockerfile-7cc2b9faed9f